File: src/main/java/com/github/fligneul/debtplugin/debt/toolwindow/DebtToolWindow.java
package com.github.fligneul.debtplugin.debt.toolwindow;

import com.github.fligneul.debtplugin.debt.service.DebtService;
import com.github.fligneul.debtplugin.debt.service.DebtServiceListener;
import com.github.fligneul.debtplugin.debt.settings.DebtSettings;
import com.github.fligneul.debtplugin.debt.settings.DebtSettingsListener;
import com.github.fligneul.debtplugin.debt.toolwindow.chart.DebtChartContainer;
import com.github.fligneul.debtplugin.debt.toolwindow.relationship.RelationshipGraphPanel;
import com.github.fligneul.debtplugin.debt.toolwindow.table.DebtTableContainer;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.ui.components.JBScrollPane;

import javax.swing.JPanel;
import javax.swing.JTabbedPane;
import java.awt.BorderLayout;

public class DebtToolWindow extends JPanel {
    private static final Logger LOG = Logger.getInstance(DebtToolWindow.class);
    private final Project project;

    private DebtTableContainer debtTableContainer;
    private DebtChartContainer debtChartContainer;
    private RelationshipGraphPanel relationshipGraphPanel;

    public DebtToolWindow(Project project) {
        this.project = project;

        createContent();

        // Refresh the table automatically when settings change (e.g., username updated)
        project.getMessageBus().connect().subscribe(DebtSettings.TOPIC, new DebtSettingsListener() {
            @Override
            public void settingsChanged(DebtSettings.State settings) {
                LOG.info("Settings changed: username=" + settings.getUsername() + " relDebtPath=" + settings.getDebtFilePath(project));
                // Apply column visibility from settings
                // Refresh data
                update();
            }
        });

        project.getMessageBus().connect().subscribe(DebtService.TOPIC, new DebtServiceListener() {
            @Override
            public void refresh() {
                if (LOG.isDebugEnabled()) LOG.debug("Refresh requested from toolwindow");
                update();
            }
        });
    }

    private void createContent() {
        // Build the existing list UI into its own panel
        debtTableContainer = new DebtTableContainer(project, new BorderLayout());

        // Build the chart tab with its own filters
        debtChartContainer = new DebtChartContainer(project, new BorderLayout());
        relationshipGraphPanel = new RelationshipGraphPanel(project);

        // Root with tabs
        final JTabbedPane tabs = new JTabbedPane();
        tabs.addTab("debts", debtTableContainer);
        tabs.addTab("modules", debtChartContainer);
        tabs.addTab("RelationShip", new JBScrollPane(relationshipGraphPanel));

        this.setLayout(new BorderLayout());
        this.add(tabs, BorderLayout.CENTER);

        update();
    }

    private void update() {
        debtTableContainer.updateTable();
        debtChartContainer.updateChart();
        relationshipGraphPanel.update();
    }

}

File: src/main/kotlin/com/github/fligneul/debtplugin/debt/DebtToolWindow.kt
package com.github.fligneul.debtplugin.debt

import com.github.fligneul.debtplugin.settings.DebtSettings
import com.github.fligneul.debtplugin.settings.DebtSettingsListener
import com.intellij.openapi.components.service
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.OpenFileDescriptor
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.ui.components.JBScrollPane
import com.intellij.ui.table.JBTable
import java.awt.BorderLayout
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.DefaultCellEditor
import javax.swing.JComboBox
import javax.swing.JButton
import javax.swing.JPanel
import javax.swing.table.DefaultTableModel

class DebtToolWindow(private val project: Project) {

    private val debtService = project.service<DebtService>()
    private val tableModel = DebtTableModel(debtService)
    private val table = JBTable(tableModel)

    init {
        val complexityComboBox = JComboBox(Complexity.values())
        table.columnModel.getColumn(6).cellEditor = DefaultCellEditor(complexityComboBox)

        val statusComboBox = JComboBox(Status.values())
        table.columnModel.getColumn(7).cellEditor = DefaultCellEditor(statusComboBox)

        val priorityComboBox = JComboBox(Priority.values())
        table.columnModel.getColumn(8).cellEditor = DefaultCellEditor(priorityComboBox)

        val riskComboBox = JComboBox(Risk.values())
        table.columnModel.getColumn(9).cellEditor = DefaultCellEditor(riskComboBox)

        // Delete button column is at index 12
        table.columnModel.getColumn(12).cellRenderer = DeleteButtonCell { row ->
            val debtItem = tableModel.debtItems[row]
            debtService.remove(debtItem)
            updateTable()
        }
        table.columnModel.getColumn(12).cellEditor = DeleteButtonCell { row ->
            val debtItem = tableModel.debtItems[row]
            debtService.remove(debtItem)
            updateTable()
        }
        table.columnModel.getColumn(12).preferredWidth = 30
        table.columnModel.getColumn(12).maxWidth = 30
        table.columnModel.getColumn(12).minWidth = 30

        // Refresh the table automatically when settings change (e.g., username updated)
        project.messageBus.connect().subscribe(DebtSettings.TOPIC, object : DebtSettingsListener {
            override fun settingsChanged(settings: DebtSettings.State) {
                updateTable()
            }
        })

        val debtFile = LocalFileSystem.getInstance().findFileByIoFile(debtService.debtFile)
    }

    fun getContent(): JPanel {
        val panel = JPanel(BorderLayout())
        panel.add(JBScrollPane(table), BorderLayout.CENTER)

        table.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                if (e.clickCount == 2) {
                    val row = table.selectedRow
                    if (row >= 0) {
                        val file = tableModel.debtItems[row].file
                        val line = table.getValueAt(row, 1) as Int
                        val virtualFile = LocalFileSystem.getInstance().findFileByPath(file)
                        if (virtualFile != null) {
                            val descriptor = OpenFileDescriptor(project, virtualFile, line - 1, 0)
                            FileEditorManager.getInstance(project).openTextEditor(descriptor, true)
                        }
                    }
                }
            }
        })

        val refreshButton = JButton("Refresh")
        refreshButton.addActionListener {
            updateTable()
        }
        panel.add(refreshButton, BorderLayout.SOUTH)

        updateTable()

        return panel
    }

    private fun updateTable() {
        tableModel.clearAll()
        debtService.all().forEach {
            tableModel.addDebtItem(it)
        }
    }
}
